market_maker_keeper/leverjfutures_market_maker_keeper_v2.py

```diff
@@ -0,0 +1,331 @@
+        self.order_book_manager.cancel_orders_with(lambda order: self.leverj_api.cancel_order(order.order_id))
+        self.order_book_manager.enable_history_reporting(self.order_history_reporter, self.our_buy_orders,
+                                                         self.our_sell_orders)
+        self.order_book_manager.start()
+
+    def main(self):
+        with Lifecycle() as lifecycle:
+            lifecycle.initial_delay(1)
+            lifecycle.on_startup(self.startup)
+            lifecycle.every(1, self.synchronize_orders)
+            lifecycle.on_shutdown(self.shutdown)
+
+    def startup(self):
+        quote_increment = self.leverj_api.get_tickSize(self.pair())
+        self.precision = -(int(log10(float(quote_increment)))+1)
+
+    def shutdown(self):
+        self.order_book_manager.cancel_all_orders()
+
+    def pair(self):
+        name_to_id_map = {'BTCDAI': '1', 'ETHDAI': '2'}
+        return name_to_id_map[self.arguments.pair.upper()]
+
+    def leverage(self):
+        leverage_in_float = self.arguments.leverage
+        if leverage_in_float is None:
+            leverage_in_float = 1.0
+        return leverage_in_float
+
+    def token_sell(self) -> str:
+        return self.arguments.pair.upper()[:3]
+
+    def token_buy(self) -> str:
+        return self.arguments.pair.upper()[3:]
+
+    def allocated_balance(self, token: str) -> Wad:
+        quote_asset_address = self.leverj_api.get_product(self.pair())["quote"]["address"]
+        # for perpetual contracts, the quote balance is allocated across instruments and sides to enter into trades
+        total_available = self.leverj_api.get_plasma_balance(quote_asset_address)
+        # adjust for leverage
+        total_available = Decimal(total_available)*Decimal(self.leverage())
+        self.logger.debug(f'total_available: {total_available}')
+        return self._allocate_to_pair(total_available).get(token)
+
+    def _allocate_to_pair(self, total_available):
+        # total number of instruments across which the total_available balance is distributed
+        # total_available is denominated in quote units
+        total_number_of_instruments = 1
+        
+        # there are 2 partitions for allocation per instrument
+        # the dai amount is divided in 2, one for the buy side and another for the sell side
+        number_of_partitions_for_allocation = Wad.from_number(total_number_of_instruments*2)
+        
+        
+        # buffer_adjustment_factor is a small intentional buffer to avoid allocating the maximum possible. 
+        # the allocated amount is a little smaller than the maximum possible allocation 
+        # and that is determined by the buffer_adjustment_factor
+        buffer_adjustment_factor = Wad.from_number(1.05)
+        
+        base = self.arguments.pair.upper()[:3]
+        quote = self.arguments.pair.upper()[3:]
+        target_price = self.price_feed.get_price()
```

Comment:
You might want to pass the target price calculated earlier in `synchronize_orders` through as an argument to ensure that all calculations are done using consistent parameters.

Reply:
We need a general idea of what the current price is to do a bunch of calculations. basically what's really in the wallet is just DAI, we basically use half that DAI to "estimate" how much fake Base token we have to be able to place sell orders